### PR TITLE
feat: Add native shuffle and columnar shuffle

### LINKED ISSUE #29: Add Shuffle support
### What is the problem the feature request solves?

Open sourcing shuffle features to Comet

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #29.

## Rationale for this change

As a columnar execution engine plugin to Spark, Comet comes out columnar operation support by providing corresponding operators to replace row-based operators in Spark. Shuffle is also a row-based operation in Spark. It is happened between the boundary of SQL operators which need to exchange data according to some specified distribution requirements. Without columnar shuffle, it means we need to do columnar to row/row to columnar around each shuffle operations. Thus, we propose Comet shuffle operators in this patch.

## What changes are included in this PR?

Two kind of shuffle operators are included in this patch: native shuffle and columnar shuffle. Both shuffle operators are columnar-based operations and use same native implementation to write shuffle data into disk. Native shuffle takes columnar batches output from Comet operators directly. Columnar shuffle takes row outputs from downstream operators which could be Spark operators or Comet operators wrapped by ColumnarToRow operator. These rows are converted into columnar batches in the native writer and written into disk.

### CODE REVIEW

**Comment by viirya** [comment]
The CI failure is:

```
Files with unapproved licenses:
  .github/pull_request_template.md
```

It will be fixed at #32 .

**Comment by viirya** [comment]
cc @sunchao

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 1]
```diff
@@ -119,6 +119,110 @@ object CometConf {
       .booleanConf
       .createWithDefault(false)
 
+  val COMET_EXEC_SHUFFLE_ENABLED: ConfigEntry[Boolean] = conf(
+    s"$COMET_EXEC_CONFIG_PREFIX.shuffle.enabled")
+    .doc("Whether to enable Comet native shuffle. By default, this config is false. " +
+      "Note that this requires setting 'spark.shuffle.manager' to" +
+      "'org.apache.spark.sql.comet.execution.CometShuffleManager'. 'spark.shuffle.manager' must " +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:126]: I think this should be `org.apache.spark.sql.comet.execution.shuffle.CometShuffleManager`

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 2]
```diff
@@ -119,6 +119,110 @@ object CometConf {
       .booleanConf
       .createWithDefault(false)
 
+  val COMET_EXEC_SHUFFLE_ENABLED: ConfigEntry[Boolean] = conf(
+    s"$COMET_EXEC_CONFIG_PREFIX.shuffle.enabled")
+    .doc("Whether to enable Comet native shuffle. By default, this config is false. " +
+      "Note that this requires setting 'spark.shuffle.manager' to" +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:125]: nit: space at the end

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 3]
```diff
@@ -119,6 +119,110 @@ object CometConf {
...
+      "'org.apache.spark.sql.comet.execution.CometShuffleManager'. 'spark.shuffle.manager' must " +
+      "be set before starting the Spark application and cannot be changed during the application")
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:127]: nit: dot at the end.

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 4]
```diff
@@ -119,6 +119,110 @@ object CometConf {
...
+    s"$COMET_EXEC_CONFIG_PREFIX.shuffle.codec")
+    .doc(
+      "The codec of Comet native shuffle used to compress shuffle data. Only zstd is supported.")
+    .stringConf
+    .createWithDefault("zstd")
+
+  val COMET_COLUMNAR_SHUFFLE_ASYNC_ENABLED: ConfigEntry[Boolean] = conf(
+    "spark.comet.columnar.shuffle.async.enabled")
+    .doc(
+      "Whether to enable asynchronous shuffle for Arrow-based shuffle. By default, this config " +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:150]: nit: maybe change "Arrow-based shuffle" to "columnar shuffle"?

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 5]
```diff
@@ -119,6 +119,110 @@ object CometConf {
...
+        "improve performance, and should be set based on the number of cores available.")
+      .intConf
+      .createWithDefault(3)
+
+  val COMET_EXEC_SHUFFLE_ASYNC_MAX_THREAD_NUM: ConfigEntry[Int] = {
+    conf("spark.comet.columnar.shuffle.async.max.thread.num")
+      .doc("Maximum number of threads on an executor used for Comet async columnar shuffle. " +
+        "By default, this config is 100. This is the upper bound of total number of shuffle " +
+        "threads per executor. In other words, if the number of cores * the number of shuffle " +
+        "threads per task `spark.comet.columnar.shuffle.async.thread.num` is larger than " +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:169]: There are two "than"s. I think this sentence is broken.

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 6]
```diff
@@ -119,6 +119,110 @@ object CometConf {
...
+      "Note that this requires setting 'spark.shuffle.manager' to" +
+      "'org.apache.spark.sql.comet.execution.CometShuffleManager'. 'spark.shuffle.manager' must " +
+      "be set before starting the Spark application and cannot be changed during the application")
+    .booleanConf
+    .createWithDefault(false)
+
+  val COMET_COLUMNAR_SHUFFLE_ENABLED: ConfigEntry[Boolean] = conf(
+    "spark.comet.columnar.shuffle.enabled")
+    .doc(
+      "Force Comet to only use Arrow-based shuffle for CometScan and Spark regular operators. " +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:134]: ditto: "columnar shuffle" instead of "Arrow-based shuffle" to be consistent with other places.

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 7]
```diff
@@ -119,6 +119,110 @@ object CometConf {
...
+        "than this config. Comet will use this config as the number of shuffle threads per " +
+        "executor instead.")
+      .intConf
+      .createWithDefault(100)
+  }
+
+  val COMET_EXEC_SHUFFLE_SPILL_THRESHOLD: ConfigEntry[Int] =
+    conf("spark.comet.columnar.shuffle.spill.threshold")
+      .doc(
+        "Number of rows to be spilled used for Comet columnar shuffle. " +
```
Comment (sunchao) [inline common/src/main/scala/org/apache/comet/CometConf.scala:179]: I think we can move this to the same line above with ".doc" but very small nit

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+import org.apache.comet.serde.QueryPlanSerde.{exprToProto, serializeDataType}
+
+/**
+ * Comet physical plan node for Spark `TakeOrderedAndProjectExec`.
+ *
+ * It is used to execute a `TakeOrderedAndProjectExec` physical operator by using Comet native
+ * engine. It is not like other physical plan nodes which are wrapped by `CometExec`, because it
+ * contains two native executions separated by a Comet shuffle exchange.
+ */
+case class CometTakeOrderedAndProjectExec(
```
Comment (sunchao) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala:45]: Looks unrelated 😂 ?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 9]
```diff
@@ -288,14 +315,67 @@ class CometSparkSessionExtensions
...
+              // Switch to use Decimal128 regardless of precision, since Arrow native execution
+              // doesn't support Decimal32 and Decimal64 yet.
+              conf.setConfString(CometConf.COMET_USE_DECIMAL_128.key, "true")
+              val cometOp = CometShuffleExchangeExec(s, shuffleType = CometNativeShuffle)
+              CometSinkPlaceHolder(nativeOp, s, cometOp)
+            case None =>
+              s
+          }
+
+        // Arrow shuffle for regular Spark operators (not Comet) and Comet operators
```
Comment (sunchao) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:360]: "Arrow shuffle" -> "Columnar shuffle"

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 10]
```diff
@@ -125,6 +128,30 @@ class CometSparkSessionExtensions
...
+            if isCometPlan(s.child) && !isCometColumnarShuffleEnabled(conf) &&
+              QueryPlanSerde.supportPartitioning(s.child.output, s.outputPartitioning) =>
+          logInfo("Comet extension enabled for Native Shuffle")
+
+          // Switch to use Decimal128 regardless of precision, since Arrow native execution
+          // doesn't support Decimal32 and Decimal64 yet.
+          conf.setConfString(CometConf.COMET_USE_DECIMAL_128.key, "true")
+          CometShuffleExchangeExec(s, shuffleType = CometNativeShuffle)
+
+        // Arrow shuffle for regular Spark operators (not Comet) and Comet operators (if configured)
```
Comment (sunchao) [inline spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:143]: ditto

**Review by sunchao — COMMENTED** [review]
Looks mostly good. The code has been reviewed internally before.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala (hunk 8 above)
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala:45]: Oh, because `CometTakeOrderedAndProjectExec` uses shuffle internally, without it we cannot add it to the initial open sourced code. So it is coupled with shuffle code in the removal list. That's why I put them back together. It can be removed and add separately, though. Let me do it to make it clearly.

**Comment by sunchao** [comment]
Let's also put some details in the PR description.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala (hunk 8 above)
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala]: Removed `CometTakeOrderedAndProjectExec` and related tests from this patch. I will add them in a separate PR.

File: common/src/main/scala/org/apache/comet/CometConf.scala (hunk 1 above)
Comment (viirya) [reply common/src/main/scala/org/apache/comet/CometConf.scala]: Nice catch!

**Comment by viirya** [comment]
Added some description there.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Merged. Thanks.
